pyfr/integrators/writers.py

```diff
@@ -76,28 +76,90 @@ def __init__(self, *args, **kwargs):
-                prank = self.rallocs.mprankmap[mrank]
-                for tag, (etype, dims) in enumerate(meleinfo):
-                    name = self._get_name_for_soln(etype, prank)
-
-                    if mrank == root:
-                        loc_names.append(name)
-                    else:
-                        rbuf = np.empty(dims, dtype=self.backend.fpdtype)
-                        rreq = comm.Recv_init(rbuf, mrank, tag)
-
-                        mpi_rbufs.append(rbuf)
-                        mpi_rreqs.append(rreq)
-                        mpi_names.append(name)
+        # Deciding if parallel
+        parallel = h5py.get_config().mpi
+        parallel &= h5py.version.version_tuple[:2] >= (2, 5)
+        parallel &= not self.cfg.getbool('soln-output', 'serial-h5', False)
+
+        if parallel:
+            self._write = self._writeParallel
+        else:
+            self._write = self._writeSerial
+
+        if parallel:
+            if rank == root:
+                solList = []
+                for mrank, meleinfo in enumerate(eleinfo):
+                    prank = self.rallocs.mprankmap[mrank]
+                    solList.extend(
+                                (self._get_name_for_soln(etype, prank), dims)
+                                for (etype, dims) in meleinfo
+                            )
+            else:
+                solList = None
+            self.solList = comm.bcast(solList, root=root)
+
+        else:
+            if rank == root:
+                self._mpi_rbufs = mpi_rbufs = []
+                self._mpi_rreqs = mpi_rreqs = []
+                self._mpi_names = mpi_names = []
+                self._loc_names = loc_names = []
+
+                for mrank, meleinfo in enumerate(eleinfo):
+                    prank = self.rallocs.mprankmap[mrank]
+                    for tag, (etype, dims) in enumerate(meleinfo):
+                        name = self._get_name_for_soln(etype, prank)
+
+                        if mrank == root:
+                            loc_names.append(name)
+                        else:
+                            rbuf = np.empty(dims, dtype=self.backend.fpdtype)
+                            rreq = comm.Recv_init(rbuf, mrank, tag)
+
+                            mpi_rbufs.append(rbuf)
+                            mpi_rreqs.append(rreq)
+                            mpi_names.append(name)
 
     def _write(self, path, solnmap, metadata):
+        raise Exception
+
+    def _writeParallel(self, path, solnmap, metadata):
+        from mpi4py import MPI
+
+        comm, rank, root = get_comm_rank_root()
+
+        h5File = h5py.File(path, 'w', driver='mpio', comm=comm)
+        sMap = {}
+        for s, shape in self.solList:
+            sMap[s] = h5File.create_dataset(s, shape, dtype=self.backend.fpdtype)
+
+        for e in solnmap:
+            s = self._get_name_for_soln(e, self.rallocs.prank)
+            sMap[s][:] = solnmap[e]
+
+        if rank == root:
+            mMap = []
```

Comment:
The very first, unreleased, version of PyFR did exactly this.  Take each metadata time, call `pickle.dumps(...)` on it and save this.  There were two reasons for moving away from this to our current format.  The first was to make it easier (longer term) for other people to poke around our file format.  Pickle is Python specific.  A string-ified Inifile is easier for people to interact with.  The second reason is left as an exercise for the reader, but is relevant for PyFR in the cloud.